### ISSUE TITLE
Help: CORS Issue

### ISSUE THREAD

**Issue by ciphercodes** [strapi/strapi#5548]
**Issue**
Hello,
I am getting my requests blocked by CORS when I try to log in to my application.

Error:

> Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at http://localhost:1337/auth/local. (Reason: CORS request did not succeed).

I am running angular for UI at http://localhost:4200 and Strapi on backend: http://localhost:1337.
This app works fine on my system but when I deploy it on cloud(development mode) and try to access it with a public IP I am running in to CORS issue.

**Steps to reproduce the behavior**

1.     Start Strapi development mode
2.     Start your Angular UI on same server
3.     API calls to Strapi via Angular are blocked in browser by CORS

**Expected behavior**
After enabling CORS in Strapi and adding headers on Angular and or nginx the request should not be blocked via CORS.

**Code snippets**

Security.json settings:
  
```json
"cors": {
    "enabled": true,
    "origin": "*",
    "expose": [
      "WWW-Authenticate",
      "Server-Authorization",
	  "Access-Control-Expose-Headers"
	    ],
    "maxAge": 31536000,
    "credentials": true,
    "methods": [
      "GET",
      "POST",
      "PUT",
      "PATCH",
      "DELETE",
      "OPTIONS",
      "HEAD"
    ],
    "headers": [
      "Content-Type",
      "Authorization",
      "X-Frame-Options",	  
      "Origin",
	  "Access-Control-Allow-Headers"
    ]
  }
```

I have the following settings on my angular app:
```javascript
     getHeaders() {
         let headers = new HttpHeaders();
         headers = headers.append("Content-Type", "application/json");
         headers = headers.append("Access-Control-Allow-Headers", "true");
         headers = headers.append("Access-Control-Allow-Origin", "*");
         return { headers };
     }
```

Proxy settings on Angular:

```json
{
  "/*": {
    "target": "http://localhost:1337",
    "secure": false
  }
}
```

After setting up nginx when I run test via CURL I get the following response:

```
server {
    listen 80;
    server_name api.test.com;

location / {

        set $cors '';
        if ($http_origin ~ '^https?://(localhost|www\.test.com\.com|portal\.test\.com)') {
                set $cors 'true';
        }

        if ($cors = 'true') {
                add_header 'Access-Control-Allow-Origin' "*" always;
                add_header 'Access-Control-Allow-Credentials' 'true' always;
                add_header 'Access-Control-Allow-Methods' 'GET, POST, PUT, DELETE, OPTIONS' always;
                add_header 'Access-Control-Allow-Headers' 'Accept,Authorization,Cache-Control,Content-Type,DNT,If-Modified-Since,Keep-Alive,Origin,User-Agent,X-Requested-With' always;
        }

        if ($request_method = 'OPTIONS') {
                add_header 'Access-Control-Max-Age' 1728000;
                add_header 'Content-Type' 'text/plain charset=UTF-8';
                add_header 'Content-Length' 0;
                return 204;
        }
        proxy_pass http://localhost:1337;
    }
}
```

Curl Request for testing:

```
curl -H "Access-Control-Request-Method: POST" -H "Origin: http://localhost" --head http://www.mysite.com/
HTTP/1.1 200 OK
Server: nginx/1.14.2
Date: Thu, 19 Mar 2020 16:29:05 GMT
Content-Type: text/html; charset=utf-8
Content-Length: 1463
Connection: keep-alive
Vary: Origin
Access-Control-Allow-Origin: *
Access-Control-Allow-Credentials: true
Access-Control-Expose-Headers: WWW-Authenticate,Server-Authorization,Access-Control-Expose-Headers
X-Powered-By: Strapi <strapi.io>
Access-Control-Allow-Origin: *
Access-Control-Allow-Credentials: true
Access-Control-Allow-Methods: GET, POST, PUT, DELETE, OPTIONS
Access-Control-Allow-Headers: Accept,Authorization,Cache-Control,Content-Type,DNT,If-Modified-Since,Keep-Alive,Origin,User-Agent,X-Requested-With
Last-Modified: Thu, 19 Mar 2020 02:42:43 GMT
Cache-Control: max-age=60
```

**System**
- Node.js version: v12.16.1
- NPM version:6.14.2
- Strapi version: 3.0.0-beta.17.4
- Database: Mysql 8.0.19
- Operating system: Linux/Debian 10

Previous issue: 5546 was closed for not using template.

**Comment by derrickmehaffy** [comment]
Can you please test on the latest Strapi version (fresh app would be fine, just for testing)

Latest is beta.19.3

**Comment by derrickmehaffy** [comment]
Also as a side note you should not need to be setting any CORs stuff inside of nginx. Here is an example config (for `/` block):

```
    location / {
        proxy_pass http://localhost:1337;
        proxy_http_version 1.1;
        proxy_set_header X-Forwarded-Host $host;
        proxy_set_header X-Forwarded-Server $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header Host $http_host;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
        proxy_pass_request_headers on;
    }
```

**Comment by derrickmehaffy** [comment]
Likewise for angular, you should not be setting any CORs stuff from the client (it won't do anything)

Remove these:
```
         headers = headers.append("Access-Control-Allow-Headers", "true");
         headers = headers.append("Access-Control-Allow-Origin", "*");
```

CORs works only on incoming response, not sent request, Strapi won't do anything with those headers incoming, but it will respond with them.

**Comment by derrickmehaffy** [comment]
As we can see with the latest beta version, on the CORs middleware: https://github.com/strapi/strapi/blob/262d3e43645203a2ed9e2549ffa8083909c523f3/packages/strapi/lib/middlewares/cors/index.js#L8-L34

it already exposes everything it needs to, and likewise if you leave just the default setting in security.json to `"enabled": true` it will by default respond back with the origin set to whoever requested it (you can also do whitelisting)

https://github.com/strapi/strapi/blob/262d3e43645203a2ed9e2549ffa8083909c523f3/packages/strapi/lib/middlewares/cors/index.js#L37-L52

This is why I would like you to test on a fresh application, as there have been lots of changes to make CORs management easier. If it works on the fresh app, then my suggestion would be to upgrade your existing project to the latest. As the issue is not a "bug".

However if the fresh app doesn't work, more will need to be looked at to see if there is a bug.

**Comment by ciphercodes** [comment]
Thanks for the input,
Let me test with the newest version and get back to you. The reason we did not migrate yet was because of some DB and group/component changes but if this issue still persist then we will have to upgrade anyway.

**Comment by derrickmehaffy** [comment]
@ciphercodes any update on this? Just wanted to see if I could close this issue out.

**Comment by ciphercodes** [comment]
Hi Derrick,
I could not test it on latest version, I did remove the headers you had mentioned from nginx and Angular app. I am going to move the angular UI on another server and from initial tests CORS is not an issue when  Angular and Strapi are on different servers it's just that when both are on same host.
I am closing this issue, thank you for your help.

**Comment by VukMNE** [comment]
Are there any news about this? Is there a way to workaround this problem?

**Comment by derrickmehaffy** [comment]
Hi @VukMNE please open up a new discussion thread so we can properly help you with your specific environment.

**Comment by VukMNE** [comment]
Hello @derrickmehaffy , I have the exact environment as @ciphercodes  specified, and I just wanted to know if there has been any breakthrough in solving this issue.

**Comment by derrickmehaffy** [comment]
As the beta version is no longer supported, we don't support this environment.
